### ISSUE TITLE
Unable to get real ip - neither in remote_ip nor X-Forwarded-For.

### ISSUE THREAD

**Issue by smathev** [lucaslorentz/caddy-docker-proxy#406]
TL;DR - Can I get the 'real remote ip' of a client who requests a website via Caddy running in a docker container used as a reverse-proxy? Below all info I think is relevant to my attempts. Please advise if it's possible or not, and what I might have missed. Thanks in advance.

OS: Debian 11.
Docker: 
Server 20.10.18
Client: 20.10.18

I'm simply unable to get any other ip's in my caddy log than the internal docker-gateway where my Caddy-docker reverse proxy is residing. This disables my attempts to filter access based on IP addresses.

I've noticed a lot of different information on how I would get the IP-address info, but nothing seems to work nor change the output of my logfile. Some places it's mentioned as a bug/feature of the docker network overlay, other places seem to have a fix but whatever I do, I'm simply unable to get any other IP's into "remote_ip" or "X-Forwarded-For" than my reverseproxy_network gateway ip.

Relevant Debug-log:

`{"level":"debug","ts":1664357917.8730943,"logger":"http.handlers.reverse_proxy","msg":"connection closed","upstream":"172.XXX.XXX.6:1880","duration":0.002890501,"request": "remote_ip":"172.XXX.XXX.1","remote_port":"52537","proto":"HTTP/1.1","method":"GET","host":"nr.mydomain.example","uri":"/comms","headers":{"Upgrade":["websocket"],"Origin":["https://nr.mydomain.example"],"Accept-Language":["da-DK,da;q=0.9,en-US;q=0.8,en;q=0.7"],"X-Forwarded-For":["172.XXX.XXX.1"],"X-Forwarded-Host":["nr.mydomain.example"],"Pragma":["no-cache"],"Connection":["Upgrade"],"Cache-Control":["no-cache"],"Sec-Websocket-Extensions":["permessage-deflate; client_max_window_bits"],"Sec-Websocket-Version":["13"],"Accept-Encoding":["gzip, deflate, br"],"Sec-Websocket-Key":["E+WaOIQuiURDAZqjwoaEdg=="],"X-Forwarded-Proto":["https"],"Cookie":[],"User-Agent":["Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36"]},"tls": "resumed":true,"version":772,"cipher_suite":4865,"proto":"http/1.1","server_name":"nr.mydomain.example"}},"duration":75.800531291}
`

My docker-compose file:
```
version: "3.7"
services:

  caddy:
    container_name: caddy_reverse_proxy
    image: caddy_test:latest
    ports:
      - target: 80
        published: 5050
        #protocol: tcp
        mode: host
      - target: 443
        published: 8443
        #protocol: tcp
        mode: host
      - target: 80
        published: 80
        #protocol: tcp
        mode: host
      - target: 443
        published: 443
        #protocol: tcp
        mode: host
    networks:
      reverseproxy_network:
        ipv4_address: 172.XXX.XXX.25
    deploy:
      mode: global
      placement:
        constraints:
          - node.role == manager
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /home/core/docker_volumes/caddy:/data
    environment:
      - CADDY_INGRESS_NETWORKS=reverseproxy_network
      - CADDY_DOCKER_CADDYFILE_PATH=/data/Caddyfile
    labels:
      caddy.email: XXX_myemail@googlemailprovider.com


networks:
  reverseproxy_network:
    external: true

```

My Dockerfile to build the caddy_test:latest image:
```
ARG CADDY_VERSION=2.6.1
FROM caddy:${CADDY_VERSION}-builder AS builder

RUN xcaddy build \
    --with github.com/lucaslorentz/caddy-docker-proxy/v2 \
    --with github.com/greenpau/caddy-security 

FROM caddy:${CADDY_VERSION}-alpine
RUN apk add --no-cache tzdata

COPY --from=builder /usr/bin/caddy /usr/bin/caddy
CMD ["caddy", "docker-proxy"]
```

One of my docker_compose files for a website I would like to serve:
```
version: "3.4"
networks:
    homeautomation_network:
        external: true 
    reverseproxy_network: 
        external: true
    nodered:
        image: nodered/node-red
        container_name: nodered
        restart: unless-stopped
        networks:
          - homeautomation_network
          - reverseproxy_network
        depends_on:
          - home-assistant
        user: "1000"
        volumes:
            - /home/core/docker_volumes/nodered/:/data:rw 
            - /etc/localtime:/etc/localtime:ro
            - /home/core/docker_volumes/homeassistant:/config:rw
        environment:
            - TZ=Europe/Copenhagen
        labels:
          caddy: nr.mydomain.example
          caddy.@denied.not: remote_ip forwarded 10.XXX.XXX.0/24 
          caddy.abort: "@denied"
          caddy.reverse_proxy: "{{upstreams 1880}}"
          caddy.authorize: with admin_policy
```

When I write `caddy.@denied.not: remote_ip forwarded 10.XXX.XXX.0/24 ` I can't access it from anywhere (including my computer on that subnet. When I add `172.XXX.XXX.0/24` I'm able to access it locally, but I'm also able to access it from the internet. And, with all request in the debug log I get the the IP of the gateway for the reverseproxy_network (IE. `172.XXX.XXX.1`)

I know it,'s the gateway, since that's the IP of the gateway when I do a `docker inspect caddy_reverse_proxy`

Is it simply that what I'm trying to do impossible, or am I missing a step which is obvious?

I'm not sure, as I'm reading information pointing both ways (possible/impossible) in a lot of different places, but none of the solutions seem to work for me.

Thanks in advance for any help - and thank you for a piece of software that seems more powerful and simpler to configure than nginx - if only I could get the 'real ip' somehow :)

**Comment by lucaslorentz** [comment]
I had it in the past, but I had proxy protocol in the load balancer receiving the request in AWS.

It looks like you're in the right path using global deploy. Same as I wrote [here](https://github.com/lucaslorentz/caddy-docker-proxy/issues/135#issuecomment-607083050).

I suppose you tried to access it on ports 5050 or 8443, right? I would expect those ports to land directly In container, skipping any caddy network gateways.

**Comment by smathev** [comment]
Thanks for getting back to me. I'm certain the 'global deploy' suggestion was from your post.

I'm fairly certain I tried both ways, IE. port 80/443 and 5050/8443. 

The reason I originally used (and still use) two different port schemes is because before Caddy I was using two different reverse proxies. One for external services, where WAN-connection was directed to port 5050/8443 on my server (to connect to the "external proxy") and another proxy on 80/443 to connect to my "local proxy" for local services.

My idea was to combine both proxies into one caddy-docker, where I limited access to "local-only" services via IP-limiting, but not getting the correct IP's meant I still have to use two proxies.

The reason why I'm opening both 80/443 and 5050/8443 was to access caddy both internally and externally. But, testing both externally (a phone connected via mobile-data and connecting to the external domain) and internally (a computer within the ethernet-network connecting to the external domain) I was still only getting the gateway-IP. 

I've been running caddy as my internal proxy for now, but whenever I use it I still only get the gateway-IP's, both in remote_ip and X-forwarded-for. This is running the proxy on port 80/443, and using only internal DNS-names. Which are served via my piHole instance.

I'm reading your response as saying that it should be possible to actually get the correct X-forwarded-for IP despite using caddy in a docker?

**Comment by francislavoie** [comment]
Check if you have Docker's userland proxy intercepting TCP packets. I didn't read the post fully but that's a common reason for not seeing the real client IP as the remote address.

**Comment by lucaslorentz** [comment]
@smathev Do you want to get it working in a Docker Swarm cluster? Or on a single machine?

I did some tests, but I'm running Docker for Mac, so my setup is a bit different.

Looks like the only way to avoid any Docker GW hop is to put CDP **only** in the host network.
If you put a CDP in any other networks, requests will do a hop in that network gateway (even bridge networks cause this).

The upstream containers can be put in the bridge network, and they would still be reachable by CDP. Putting all of them in the host network would also work, but it would be a nightmare to manage all ports without conflicts.

CDP will always try to match container networks with it's own networks, and it will never match with this setup. Therefore we need to inform CDP that it should use bridge network to reach other containers: `CADDY_INGRESS_NETWORKS=bridge`.

It's important as well to add labels outside the "deploy" section. Otherwise, CDP will check docker service task networks (which are empty), instead of container networks (which have the bridge network).

Compose file working for me:
```
version: "3.7"

services:
  caddy:
    image: lucaslorentz/caddy-docker-proxy:ci-alpine
    ports:
      - target: 80
        published: 80
        mode: host
    networks:
      - host
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - CADDY_INGRESS_NETWORKS=bridge
    deploy:
      mode: global
      placement:
        constraints:
          - node.role == manager
      labels:
        caddy: "http://*"
        caddy.respond: "/whatsmyip \"Hello {http.request.remote.host}\""

  backend:
    image: sixeyed/whoami-dotnet:3.0-master-stretch-arm64
    networks:
      - bridge
    environment:
      - ASPNETCORE_URLS=http://+:8901
    labels:
      caddy: "http://*"
      caddy.reverse_proxy: "{{upstreams 8901}}"
    deploy:
      mode: global
      placement:
        constraints:
          - node.role == manager

volumes:
  caddy_data: {}

networks:
  host:
    external: true
    name: host
  bridge:
    external: true
    name: bridge
```

Test ip address:
```
curl http://localhost/whatsmyip
```

Test upstream reachability:
```
curl http://localhost
```

Feel free to change backend image, I'm testing on an M1 Mac and `sixeyed/whoami-dotnet:3.0-master-stretch-arm64` was the only one I found for arm64.

**Comment by lucaslorentz** [comment]
Closing because it worked with configs above.
